### PR TITLE
[DNM: Draft][SILOpt] TempRValueElimination: 'let' copy source values cant be over…

### PR DESCRIPTION
**Pull request by hyp**
…writen by apply instructions

TempRValueElimination checks if a copy of a value can be eliminated. Checking that the original source value was not modified during the copy is one of the requirements for the elimination. Prior to this change, TempRValueElimination checked if apply instructions could've potentially modified the original source value, even if the source value was an immutable 'let'. This prevented some optimization opportunities. With this change, we can recognize that 'let' source value won't be modified by an apply, so that we can proceed with copy elimination.

### CODE REVIEW

**Comment by hyp** [comment]
still needs tests

**Comment by hyp** [comment]
@swift-ci please benchmark

**Comment by hyp** [comment]
@swift-ci please test

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp [hunk 1]
```diff
@@ -357,8 +357,16 @@ SILInstruction *TempRValueOptPass::getLastUseWhileSourceIsNotModified(
...
+        SILValue object = getUnderlyingObject(copySrc);
+        if (auto *ASI = dyn_cast<AllocStackInst>(object)) {
+          if (ASI->isLet())
+            return true;
+        }
+        return false;
+      };
       if ((FullApplySite::isa(inst) || isa<YieldInst>(inst)) &&
-          aa->mayWriteToMemory(inst, copySrc)) {
+          !isLetObject(copySrc) && aa->mayWriteToMemory(inst, copySrc)) {
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/TempRValueElimination.cpp:369]: Now, you can just use `bool swift::isLetAddress(SILValue address)`. That uses the canonical AccessBase utility. It's a bit less ad-hoc and you don't need to DIY.

**Review by atrick — COMMENTED** [review]
Can you add a .sil test to temp_rvalue_opt_ossa.sil?

File: lib/SILOptimizer/Transforms/TempRValueElimination.cpp (hunk 1 above)
Comment (atrick) [reply lib/SILOptimizer/Transforms/TempRValueElimination.cpp:369]: @meg-gupta pointed out that if you just check `AllocStack::isLet()` inside `isLetAddress`, then you shouldn't need to touch this pass at all.

**Comment by hyp** [comment]
Closing in favor of https://github.com/apple/swift/pull/61504

**Comment by atrick** [comment]
I think @meg-gupta will introduce a PR very similar to this, maybe with a comment that we can't safely rely on `mayWriteToMemory` to recognize 'let' variables

**Pull request closed without merging**